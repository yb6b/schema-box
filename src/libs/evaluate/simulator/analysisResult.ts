import { comboFeelData, keyFeelData } from '../feelData'
import type { Segment } from './segmentation'

type Dist = number[]

export type FullAnalysisResult = ReturnType<AnalysisResult['reform']>
/**
 * 赛码器运行结束后的结果
 */
export class AnalysisResult {
  /**
   * 为了性能,需要提供几个最大值用于初始化,
   *
   * 新建完成后,请检查 collisionKeys 和 commitLength 参数
   */
  constructor(maxCollision: number, maxWords: number, maxCodeLength: number) {
    this.collisionDist = Array(maxCollision).fill(0)
    this.wordsDist = Array(maxWords).fill(0)
    this.codeLenDist = Array(maxCodeLength).fill(0)
    for (const k in keyFeelData)
      this.keysDist[k] = 0
    for (const k in comboFeelData)
      this.comboDist[k] = 0
  }

  reform() {
    const lackCounter = this.lackCounter
    let lacks = 0
    for (const e of lackCounter.values())
      lacks += e
    const lackString = [...lackCounter.keys()].sort().join('')

    const wordsDist = dropLastFalsyItems(this.wordsDist)
    const singleCount = wordsDist[0]
    let commit = 0
    let char = 0
    for (let i = 0; i < wordsDist.length; i++) {
      const n = wordsDist[i]
      commit += n
      char += n * (i + 1)
    }

    const collisionDist = dropLastFalsyItems(this.collisionDist)
    let collision = 0
    for (const e of collisionDist)
      collision += e

    const codeLenDist = dropLastFalsyItems(this.codeLenDist)
    let codeLen = 0
    for (let i = 0; i < codeLenDist.length; i++) {
      const e = codeLenDist[i]
      codeLen += e * (i + 1)
    }

    const keysDist = dropNanInRecord(this.keysDist)
    const kbdRowDist: Dist = [0, 0, 0, 0]
    const finDist: Dist = Array(11).fill(0)
    let keys = 0
    for (const [k, n] of Object.entries(keysDist)) {
      keys += n
      const row = keyFeelData[k].row
      kbdRowDist[row] += n
      const fin = keyFeelData[k].fin
      finDist[fin] += n
    }

    const comboDist = dropNanInRecord(this.comboDist)
    let combo = 0
    let Eq = 0
    let double = 0
    let singleSpan = 0
    let multiSpan = 0
    let longFD = 0
    let littleFD = 0
    let sameFingers = 0
    let leftLeft = 0
    let leftRight = 0
    let rightRight = 0
    let rightLeft = 0
    // TODO 处理大写字母
    for (const [k, n] of Object.entries(comboDist)) {
      combo += n
      const feeldata = comboFeelData[k]
      if (!feeldata)
        continue
      Eq += feeldata.eq * n
      double += (feeldata.dc as unknown as number) * n
      singleSpan += (feeldata.ss as unknown as number) * n
      multiSpan += (feeldata.ms as unknown as number) * n
      longFD += (feeldata.lfd as unknown as number) * n
      littleFD += (feeldata.pd as unknown as number) * n
      const fin1 = keyFeelData[k[0]].fin
      const fin2 = keyFeelData[k[1]].fin
      sameFingers += (fin1 === fin2) as unknown as number * n
      const right1 = fin1 > 5
      const right2 = fin2 > 5
      leftLeft += (!right1 && !right2) as unknown as number * n
      leftRight += (!right1 && right2) as unknown as number * n
      rightRight += (right1 && right2) as unknown as number * n
      rightLeft += (right1 && right2) as unknown as number * n
    }
    Eq = (Eq * 10 | 0) / 10
    const diffHand = leftRight + rightLeft

    return {
      lacks,
      lackString,
      lackCounter,
      wordsDist,
      singleCount,
      commit,
      char,
      collision,
      collisionDist,
      codeLen,
      codeLenDist,
      keys,
      keysDist,
      kbdRowDist,
      finDist,
      comboDist,
      combo,
      Eq,
      double,
      singleSpan,
      multiSpan,
      longFD,
      littleFD,
      sameFingers,
      leftLeft,
      leftRight,
      rightRight,
      rightLeft,
      diffHand,
    } as const
  }

  /** 选重按键, 如果选重超出了,会翻页 */
  collisionKeys = ' 23456789'
  /** 上屏码长 */
  commitLength = 4

  /** 所有按键组合的为了性能，一开始只把键位组合统计到map里 */
  private comboDist: Record<string, number> = {}
  /** 方案里缺少的字符, 无法打出的字符 */
  private lackCounter = new Map<string, number>()

  /**
   * 各长度的词语的次数,
   * 例：索引0表示上屏单字的次数
   *
   * 需要提供最大值,用于性能优化
   */
  private wordsDist: Dist

  /**
   * 选重频数分布
   *
   * 需要提供最大值,用于性能优化
   *
   * - 索引 0 顶字上屏
   * - 索引 1 首选
   * - 索引 2 二重
   */
  private collisionDist: Dist

  /**
   * 不同编码长度的分布,例：索引0表示1码长的上屏次数
   *
   * 需要提供最大值,用于性能优化
   */
  private codeLenDist: Dist

  /** 各按键的频数 */
  private keysDist: Record<string, number> = {}

  /** these codes  */
  private _t = ''
  /** last code */
  private _l = ''

  /**
   * 添加分词结果, 初步分析。
   */
  add(aSegValue: Segment) {
    this.handle(aSegValue)
    this._l = this._t[this._t.length - 1]
  }

  private handle(value: Segment) {
    switch (value.type) {
      // 词条
      case 1:{
        this.hTT(value)
        // 只有字典里的词条才计算组合手感
        this.hCombo()
        break
      }
      // 缺字
      case 5:{
        this.hLack(value)
        // 缺字情况下，没有编码
        return
      }
    }
    // 标点和英文
    this.hKeys()
  }

  /** 处理缺字 */
  private hLack(value: Segment) {
    const w = this.lackCounter.get(value.bw!)
    this.lackCounter.set(value.bw!, w ? w + 1 : 1)
    this._l = ''
    this._t = ''
  }

  /** 处理字典树匹配到的情况 */
  private hTT(value: Segment) {
    let code = value.ttv!.i![1]

    const collision = value.ttv!.c
    this.collisionDist[collision - 1]++

    const wordsLen = [...value.ttv!.i![0]].length
    this.wordsDist[wordsLen - 1]++

    // 码长不考虑选重键
    this.codeLenDist[code.length - 1]++

    // 要加选重键的情况
    if (code.length < this.commitLength || collision !== 1)
      code += this.collisionKeys[collision - 1]
    this._t = code
  }

  /** 处理按键 */
  private hKeys() {
    for (const i of this._t)
      this.keysDist[i]++
  }

  /** 分析按键组合 */
  private hCombo() {
    // 两键组合
    const expandCode = this._l + this._t
    for (let i = 1; i < expandCode.length; i++) {
      const comb = expandCode[i - 1] + expandCode[i]
      this.comboDist[comb]++
    }
  }
}

function dropLastFalsyItems(d: Dist) {
  let lastIndex = 0
  for (let i = d.length - 1; i >= 0; i--) {
    if (d[i]) {
      lastIndex = i + 1
      break
    }
  }
  return d.slice(0, lastIndex)
}

function dropNanInRecord(d: Record<string, number>) {
  return Object.fromEntries(Object.entries(d).filter(([_, n]) => !!n))
}
