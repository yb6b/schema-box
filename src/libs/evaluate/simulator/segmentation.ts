/**
 * @file 按照字典树，拆分文章成一条条词语
 * @author yb6b
 */

import { PUNCTUATIONS } from '../../schema/keys'
import type { TrieTreeValue, UniqueTrieTree } from './uniqueTrieTree'

/**
 * 按照字典树切文章，要预处理文章中的控制字符
 */
export class Segment {
  constructor(
    public tree: UniqueTrieTree,
    public article: string,
  ) { }

  /** 把结果转换成对象，在在matchNext之后执行 */
  collect(): SegValue {
    switch (this.type) {
      case 1:
        return { type: 1, value: this.ttv! }
      case 5:
        return { type: 5, words: this.bw! }
      default:
        return { type: this.type, words: this.bw!, code: this.bc! }
    }
  }

  index = 0
  /**
   * 匹配下一个。如果要重新开始，手动设置 this.index=0
   * @returns 如果是-1 则全文匹配结束
   */
  matchNext() {
    const i = this.index

    // 已经结尾了，跳过
    if (i >= this.article.length)
      return -1

    // 是不是字典树里的
    if (this.mTT(i))
      return this.index += this.ttv!.i![0].length

    if (this.mPunc(i) // 是不是中文标点符号
      || this.mBD(i, PUNCTUATIONS.en) // 是不是英文标点符号
      || this.mBD(i, PUNCTUATIONS.uni) // 是不是中英文通用的标点符号
    )
      return this.index += this.bw!.length

    // 都不是，那只能是缺字
    this.bw = String.fromCodePoint(this.article.codePointAt(i)!) // 要考虑双字的unicode
    this.type = 5
    return this.index += this.bw.length
  }

  /**
   * 临时切片的类型：
   * 1. 码表的字词
   * 2. 中文标点
   * 3. 英文和英文标点
   * 4. 无论中英文都能打出的标点
   * 5. 缺字
   */
  type: SegValue['type'] = 1
  /** 临时的最佳词条 */
  bw?: string
  /** 临时的最佳编码 */
  bc?: string
  /** 临时的字典树 */
  ttv?: TrieTreeValue

  /** 预处理 */
  private pre() {
    if (this.article.includes('　'/* 全角空格 */))
      this.article = this.article.replace('　', '')
  }

  /**
   * match trie tree
   * 从文章的特定索引号开始, 找到最佳匹配的字典树的词.
   * 找到的最佳匹配会放在this.ttv里
   * @return 能否匹配到
   */
  private mTT(index: number): boolean {
    const maxIndex = this.article.length
    let wd = this.article[index]
    let alreadyHadResult = false

    for (let i = index; i < maxIndex;) {
      const ttvalue = this.tree.get(wd)

      // 未收录
      if (!ttvalue)
        return alreadyHadResult

      // 收录了

      // 正好最长匹配
      if (wd.length === ttvalue.p) {
        this.ttv = ttvalue
        return true
      }
      // 不是最长的,继续探索更长的
      if (ttvalue.i) {
        this.ttv = ttvalue
        alreadyHadResult = true
        i = index + ttvalue.p
        wd = ttvalue.i[0]
        continue
      }
      // 是空穴，不可能
      throw new Error(`不可能的空穴${index} ${JSON.stringify(ttvalue)}`)
    }
    // 到了最后,却不是最长匹配
    return alreadyHadResult
  }

  /**
   * match punctuation
   * 从指定索引号开始, 找到下一个中文标点符号
   * 最佳匹配会放在this.bc this.bw里
   * @return 能不能匹配到
   */
  private mPunc(index: number): boolean {
    const dict = PUNCTUATIONS.cn
    let wd = this.article[index]
    let punc = dict[wd]
    if (!punc)
      return false
    // 处理两个字的标点
    if (punc === '…' || punc === '—') {
      if (punc === this.article[index + 1]) {
        wd += wd
        punc = dict[wd]
        if (!punc)
          return false
      }
    }
    this.bw = wd
    this.bc = punc
    return true
  }

  /**
   * match by dict
   * 用于找到下一组英文或通用符号
   * 最佳匹配会放在this.bc this.bw里
   * @return 能不能匹配到
   */
  private mBD(index: number, dict: Record<string, string>) {
    const alreadyHadResult = false
    const len = this.article.length
    let c = ''
    let w = ''
    for (let i = index; i < len; i++) {
      const wd = this.article[i]
      const code = dict[wd]
      if (!code)
        return alreadyHadResult
      c += code
      w += wd
      this.bc = c
      this.bw = w
    }
    // 到结尾都是符号
    return true
  }
}

interface SegValueCn {
  type: 1
  value: TrieTreeValue
}

interface SegValuePunc {
  type: 2
  words: string
  code: string
}

interface SegValueEn {
  type: 3
  words: string
  code: string
}

interface SegValueUni {
  type: 4
  words: string
  code: string
}

interface SegValueLack {
  type: 5
  words: string
}

export type SegValue = SegValueCn | SegValueEn | SegValueLack | SegValuePunc | SegValueUni
