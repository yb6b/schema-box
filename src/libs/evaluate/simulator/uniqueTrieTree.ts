import type { SchemaDictItem } from '../../schema/index'

/**
 * 字典树中，每个词语的值。
 */
export interface TrieTreeValue {
  /** 重码数 */
  c: number
  /**
   * 以该词为前缀的词条的最短字数，这是 `UniqueTrieTree` 维护的
   * 新建时手动设置为0
   *
   */
  p: number
  /**
   * 包含编码、行号、词条信息,
   * 若是null,表示该记录是空穴
   * 手动添加时, 请勿设置为null
   */
  i: SchemaDictItem | null
}

/** 这种字典树里每个词条只保留一个编码 */
export type UniqueTrieTree = Map<string, TrieTreeValue>

/**
 * 获取字典树的值
 * 空穴词,或没有缺失的词,返回undefined
 */
export function treeGet(tree: UniqueTrieTree, words: string) {
  const r = tree.get(words)
  if (r?.i)
    return r
  return undefined
}

/**
 * 加入一条词,请确保数据有效.
 * 词条冲突时, 保留码长更短的
 * @param tree 添加到哪个字典树里？
 * @param item 添加哪个词条？
 * @param collision 这个词条的重码是几重？
 * @return 是否正常添加进字典树
 */
export function treeAddShortest(tree: UniqueTrieTree, item: SchemaDictItem, collision: number): boolean {
  const words = item.words
  const oldValue = tree.get(words)
  // 词条已经存在
  if (oldValue && oldValue.i) {
    if (oldValue.i.code.length <= item.code.length)
      // 码长更长则跳过
      return false
    oldValue.i = item
    oldValue.c = collision
    return true
  }
  // 词条不存在, 先添加其中
  tree.set(words, { i: item, c: collision, p: 0 })
  // 添加前缀
  const wordsLength = words.length
  for (let i = 1; i < wordsLength; i++) {
    const prefixWords = words.slice(0, -i)
    const prefixValue = tree.get(prefixWords)
    // 前缀存在，修改其 prefixeeLen 值(变得更小)
    if (prefixValue) {
      prefixValue.p = wordsLength
      if (prefixValue.i)
        return true
    }
    // 前缀不存在，该前缀一定是空穴
    else {
      tree.set(prefixWords, { i: null, c: 0, p: wordsLength })
    }
  }
  return true
}

/**
 * 加入一条词,请确保数据有效.
 * 词条冲突时, 只保留后添加的. 通常也是行号靠后的
 * @param tree 添加到哪个字典树里？
 * @param item 添加哪个词条？
 * @param collision 这个词条的重码是几重？
 * @return 是否正常添加进字典树
 */
export function treeAddLatest(tree: UniqueTrieTree, item: SchemaDictItem, collision: number): boolean {
  const words = item.words
  const oldValue = tree.get(words)
  // 词条已经存在, 也不是空穴
  if (oldValue && oldValue.i) {
    // 不管如何,直接添加
    oldValue.i = item
    oldValue.c = collision
    return true
  }
  // 词条不存在, 先添加其中
  tree.set(words, { i: item, c: collision, p: 0 })
  // 添加前缀
  const wordsLength = words.length
  for (let i = 1; i < wordsLength; i++) {
    const prefixWords = words.slice(0, -i)
    const prefixValue = tree.get(prefixWords)
    // 前缀存在，修改其 prefixeeLen 值(变得更小)
    if (prefixValue) {
      prefixValue.p = wordsLength
      if (prefixValue.i)
        return true
    }
    // 前缀不存在，该前缀一定是空穴
    else {
      tree.set(prefixWords, { i: null, c: 0, p: wordsLength })
    }
  }
  return true
}

/**
 * 删除数据库里的一条词
 * @param tree 从哪个字典树里删除词条
 * @param words 词语
 * @returns 是否真地改了数据
 */
export function treeDelete(tree: UniqueTrieTree, words: string): boolean {
  const oldValue = tree.get(words)
  // 已收录的词，且不是空穴
  if (oldValue && oldValue.i) {
    // 字典树里,既有更长的词,又有更短的词
    if (oldValue.p > words.length) {
      // ① 当前词设为空穴
      oldValue.i = null
      // ② 修改短词中的 最短前缀长
      const newPrefixeeLen = oldValue.p
      for (let i = 1; i < words.length; i++) {
        const prefixWords = words.slice(0, -i)
        // 因为前缀词必然存在，所以关闭ts检查
        const prefixValue = tree.get(prefixWords)
        if (!prefixValue)
          continue
        prefixValue.p = newPrefixeeLen
        // 直到非空穴词才结束
        if (prefixValue.i)
          return true
      }
    }
    // 如果没有更长的词,删除它和它的空穴前缀
    else {
      // ① 删除当前词
      tree.delete(words)
      // ② 删除前缀
      for (let i = 1; i < words.length; i++) {
        const prefixWords = words.slice(0, -i)
        // 直到非空穴
        if (tree.get(prefixWords)?.i)
          return true
        else
          tree.delete(prefixWords)
      }
    }
    return true
  }
  // 未收录的词,或空穴,结束
  return false
}
