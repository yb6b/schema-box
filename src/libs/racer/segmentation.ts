/**
 * @file 按照字典树，拆分文章成一条条词语
 * @author yb6b
 */

import { PUNCTUATIONS } from '../schema/keys'
import type { TrieTreeValue, UniqueTrieTree } from './uniqueTrieTree'

interface SegValueCn {
  type: 'c'
  index: number
  words: string
  value: TrieTreeValue
}

interface SegValuePunc {
  type: 'p'
  index: number
  words: string
  code: string
}

interface SegValueEn {
  type: 'e'
  index: number
  words: string
  code: string
}

interface SegValueUni {
  type: 'u'
  index: number
  words: string
  code: string
}

interface SegValueLack {
  type: 'l'
  index: number
  words: string
}

export type SegValue = SegValueCn | SegValueEn | SegValueLack | SegValuePunc | SegValueUni

/**
 * 用贪婪算法拆分文章。优先用最长的词匹配。
 * @param src 要分词的文章，请确保文章不是空字符串
 * @param tree 方案的字典树
 */
export function segmentGreedily(src: string, tree: UniqueTrieTree): SegValue[] {
  const maxIndex = src.length - 1
  const result: SegValue[] = []
  let beginIndex = 0
  let endIndex = 0
  let bestWords: string
  let bestValue: TrieTreeValue

  const maybeSignalButZh = () => {
    const getCode = (dict: any) => {
      const ei = endIndex
      const bi = beginIndex
      const d = dict
      const tmpResult: string[] = new Array(ei - bi)
      for (let i = bi; i < ei; i++)
        tmpResult[i - bi] = d[src[i]]
      return tmpResult.join('')
    }
    const twoLetterPunctuation = (punc: string) => {
      const puncCode = PUNCTUATIONS.cn[punc]
      if (src[beginIndex + 1] === punc) {
        result.push({ type: 'p', words: punc + punc, index: beginIndex, code: puncCode + puncCode })
        beginIndex += 2
      }
      else {
        result.push({ type: 'p', words: punc, index: beginIndex, code: puncCode })
        beginIndex += 1
      }
    }
    const firstWord = src[beginIndex]
    endIndex = beginIndex
    // 是中文标点？
    if (firstWord in PUNCTUATIONS.cn) {
      switch (firstWord) {
        case '…': twoLetterPunctuation('…')
          break
        case '—': twoLetterPunctuation('—')
          break
        default:result.push({ type: 'p', words: firstWord, index: beginIndex, code: PUNCTUATIONS.cn[firstWord] })
          beginIndex++
          break
      }
    }
    // 中英通用的标点？
    else if (firstWord in PUNCTUATIONS.uni) {
      while (src[endIndex] in PUNCTUATIONS.uni && endIndex <= maxIndex) endIndex++
      result.push({ type: 'u', index: beginIndex, words: src.slice(beginIndex, endIndex), code: getCode(PUNCTUATIONS.uni) })
      beginIndex = endIndex
    }
    // 英文符号？
    else if (firstWord in PUNCTUATIONS.en) {
      while (src[endIndex] in PUNCTUATIONS.en && endIndex <= maxIndex) endIndex++
      result.push({ type: 'e', index: beginIndex, words: src.slice(beginIndex, endIndex), code: getCode(PUNCTUATIONS.en) })
      beginIndex = endIndex
    }
    // 只能是缺字了
    else {
      result.push({ type: 'l', words: firstWord, index: beginIndex })
      beginIndex += firstWord.length
    }
  }
  // 如果第一轮就挂了，需要best*数据回退
  const scanTree = () => {
    // 先处理第一次运行
    while (true) {
      // 到文末
      if (maxIndex < endIndex) {
        result.push({ type: 'c', index: beginIndex, words: bestWords, value: bestValue })
        beginIndex += bestWords.length
        return
      }
      const tmpWords = src.slice(beginIndex, endIndex)
      const tmpValue = tree.data.get(tmpWords)
      if (tmpValue) {
        // 到最长
        if (tmpValue.prefixeeLen === tmpWords.length) {
          result.push({ type: 'c', index: beginIndex, words: tmpWords, value: tmpValue })
          beginIndex = endIndex
          return
        }
        // 否则仍需要迭代，用几个 best* 暂存结果
        bestWords = tmpWords
        bestValue = tmpValue
        endIndex = beginIndex + tmpValue.prefixeeLen
      }
      // 找不到了
      else {
        result.push({ type: 'c', index: beginIndex, words: bestWords, value: bestValue })
        beginIndex += bestWords.length
        return
      }
    }
  }

  // 每次循环生成添加一个 Seg
  do {
    const firstWord = String.fromCodePoint(src.codePointAt(beginIndex) as number)
    const firstWordValue = tree.data.get(firstWord)
    if (firstWordValue) {
      // 数据库里只收录了该单字，不可能是空穴
      if (firstWord.length === firstWordValue.prefixeeLen) {
        result.push({ type: 'c', index: beginIndex, words: firstWord, value: firstWordValue })
        beginIndex += firstWord.length
        continue
      }

      bestValue = firstWordValue
      bestWords = firstWord
      endIndex = beginIndex + firstWordValue.prefixeeLen

      // 第一个字不是空穴
      if (firstWordValue.code) {
        scanTree()
      }
      // 第一个字是空穴
      else {
        const secondWord = src.slice(beginIndex, endIndex)
        const secondWordValue = tree.data.get(secondWord)
        // 能匹配更长的，必然不是空首字
        if (secondWordValue) {
          // 数据库里只收录了词，不可能是空穴
          if (secondWord.length === secondWordValue.prefixeeLen) {
            result.push({ type: 'c', index: beginIndex, words: secondWord, value: secondWordValue })
            beginIndex += secondWord.length
            continue
          }
          bestWords = secondWord
          bestValue = secondWordValue
          endIndex = beginIndex + secondWordValue.prefixeeLen
          scanTree()
        }
        // 首字是空的，有可能是标点等
        else {
          maybeSignalButZh()
        }
      }
    }
    // 未收录或空穴的，是缺字或别的情况
    else {
      maybeSignalButZh()
    }
  } while (beginIndex <= maxIndex)

  return result
}

export function* generateSegmentGreedily(src: string, tree: UniqueTrieTree) {
  const maxIndex = src.length - 1
  let beginIndex = 0
  let endIndex = 0
  let bestWords: string
  let bestValue: TrieTreeValue

  // 每次循环生成添加一个 Seg
  do {
    const firstWord = String.fromCodePoint(src.codePointAt(beginIndex) as number)
    const firstWordValue = tree.data.get(firstWord)
    if (firstWordValue) {
      // 数据库里只收录了该单字，不可能是空穴
      if (firstWord.length === firstWordValue.prefixeeLen) {
        yield { type: 'c', index: beginIndex, words: firstWord, value: firstWordValue } as SegValueCn
        beginIndex += firstWord.length
        continue
      }

      bestValue = firstWordValue
      bestWords = firstWord
      endIndex = beginIndex + firstWordValue.prefixeeLen

      // 第一个字不是空穴
      if (firstWordValue.code) {
        yield scanTree()
      }
      // 第一个字是空穴
      else {
        const secondWord = src.slice(beginIndex, endIndex)
        const secondWordValue = tree.data.get(secondWord)
        // 能匹配更长的，必然不是空首字
        if (secondWordValue) {
          // 数据库里只收录了词，不可能是空穴
          if (secondWord.length === secondWordValue.prefixeeLen) {
            yield { type: 'c', index: beginIndex, words: secondWord, value: secondWordValue } as SegValueCn
            beginIndex += secondWord.length
            continue
          }
          bestWords = secondWord
          bestValue = secondWordValue
          endIndex = beginIndex + secondWordValue.prefixeeLen
          yield scanTree()
        }
        // 首字是空的，有可能是标点等
        else {
          yield maybeSignalButZh()
        }
      }
    }
    // 未收录或空穴的，是缺字或别的情况
    else {
      yield maybeSignalButZh()
    }
  } while (beginIndex <= maxIndex)

  function maybeSignalButZh(): SegValue {
    const getCode = (dict: any) => {
      const ei = endIndex
      const bi = beginIndex
      const d = dict
      const tmpResult: string[] = new Array(ei - bi)
      for (let i = bi; i < ei; i++)
        tmpResult[i - bi] = d[src[i]]
      return tmpResult.join('')
    }
    const twoLetterPunctuation = (punc: string): SegValuePunc => {
      const puncCode = PUNCTUATIONS.cn[punc]
      if (src[beginIndex + 1] === punc) {
        const tmp: SegValuePunc = { type: 'p', words: punc + punc, index: beginIndex, code: puncCode + puncCode }
        beginIndex += 2
        return tmp
      }
      else {
        const tmp: SegValuePunc = { type: 'p', words: punc, index: beginIndex, code: puncCode }
        beginIndex += 1
        return tmp
      }
    }
    let tmp: SegValue
    const firstWord = src[beginIndex]
    endIndex = beginIndex
    // 是中文标点？
    if (firstWord in PUNCTUATIONS.cn) {
      switch (firstWord) {
        case '…': return twoLetterPunctuation('…')
        case '—': return twoLetterPunctuation('—')
        default: tmp = { type: 'p', words: firstWord, index: beginIndex, code: PUNCTUATIONS.cn[firstWord] }
          beginIndex++
      }
    }
    // 中英通用的标点？
    else if (firstWord in PUNCTUATIONS.uni) {
      while (src[endIndex] in PUNCTUATIONS.uni && endIndex <= maxIndex) endIndex++

      tmp = { type: 'u', index: beginIndex, words: src.slice(beginIndex, endIndex), code: getCode(PUNCTUATIONS.uni) }
      beginIndex = endIndex
    }
    // 英文符号？
    else if (firstWord in PUNCTUATIONS.en) {
      while (src[endIndex] in PUNCTUATIONS.en && endIndex <= maxIndex) endIndex++
      tmp = { type: 'e', index: beginIndex, words: src.slice(beginIndex, endIndex), code: getCode(PUNCTUATIONS.en) }
      beginIndex = endIndex
    }
    // 只能是缺字了
    else {
      tmp = { type: 'l', words: firstWord, index: beginIndex }
      beginIndex += firstWord.length
    }
    return tmp
  }

  // 如果第一轮就挂了，需要best*数据回退
  function scanTree() {
    // 先处理第一次运行
    while (true) {
      let tmp: SegValue
      // 到文末
      if (maxIndex < endIndex) {
        tmp = { type: 'c', index: beginIndex, words: bestWords, value: bestValue }
        beginIndex += bestWords.length
        return tmp
      }
      const tmpWords = src.slice(beginIndex, endIndex)
      const tmpValue = tree.data.get(tmpWords)
      if (tmpValue) {
        // 到最长
        if (tmpValue.prefixeeLen === tmpWords.length) {
          tmp = { type: 'c', index: beginIndex, words: tmpWords, value: tmpValue }
          beginIndex = endIndex
          return tmp
        }
        // 否则仍需要迭代，用几个 best* 暂存结果
        bestWords = tmpWords
        bestValue = tmpValue
        endIndex = beginIndex + tmpValue.prefixeeLen
      }
      // 找不到了
      else {
        tmp = { type: 'c', index: beginIndex, words: bestWords, value: bestValue }
        beginIndex += bestWords.length
        return tmp
      }
    }
  }
}
