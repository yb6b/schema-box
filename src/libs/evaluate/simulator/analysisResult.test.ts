import { bench } from 'vitest'
import type { SchemaDictItem } from 'libs/schema'
import { AnalysisResult } from './analysisResult'
import { createTree, treeAdd } from './uniqueTrieTree'
import { Segment } from './segmentation'

const v: SchemaDictItem[] = [
  ['不', 'aa', 1],
  ['不是是', 'aa bB', 2],
  ['是', 'b', 3],
  ['$', 's', 4],
  ['￥', 's', 5],
  ['\n', 's', 5],
]

const tree = createTree()
v.forEach(v => treeAdd(tree, v, 1))
const seg = new Segment(tree, '不是，也$是不是是')
// aa b3,?s4b3aa bB2
v.forEach(v => treeAdd(tree, v as any, 1))

it('正常运行', () => {
  const r = run()
  expect(r).toMatchSnapshot()
})

function run() {
  seg.index = 0
  const ar = new AnalysisResult(3, 3, 5)
  let i = seg.next()
  while (i !== -1) {
    ar.add(seg)
    i = seg.next()
  }
  ar.reform()
  return ar
}
