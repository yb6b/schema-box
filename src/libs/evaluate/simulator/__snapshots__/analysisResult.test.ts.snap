// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`正常运行 1`] = `
AnalysisResult {
  "Eq": 2.4,
  "LeftLeft": 0,
  "LeftRight": 6,
  "RightLeft": 7,
  "RightRight": 2,
  "_Eq10": 24,
  "_hp": [
    0,
    6,
    7,
    2,
  ],
  "_l": " ",
  "_l2": "b",
  "_t": "b ",
  "codeLength": 16,
  "codeLengthDist": [
    0,
    5,
    2,
    0,
    0,
    0,
  ],
  "collision": 0,
  "collisionChar": 7,
  "collisionDist": [
    7,
    0,
    0,
  ],
  "collisionKeys": " 23456789",
  "combo": 15,
  "comboDist": Map {},
  "commit": 7,
  "commitLength": 4,
  "double": 0,
  "fingersDist": [
    8,
    4,
    1,
    0,
    5,
    0,
    0,
    0,
    0,
    0,
    0,
  ],
  "hanzi": 7,
  "kbdRowDist": [
    8,
    0,
    5,
    5,
  ],
  "keys": 36,
  "keysDist": {
    " ": 8,
    "'": 0,
    ",": 0,
    "-": 0,
    ".": 0,
    "/": 0,
    "0": 0,
    "1": 0,
    "2": 0,
    "3": 0,
    "4": 0,
    "5": 0,
    "6": 0,
    "7": 0,
    "8": 0,
    "9": 0,
    ";": 0,
    "=": 0,
    "[": 0,
    "]": 0,
    "\`": 0,
    "a": 4,
    "b": 5,
    "c": 0,
    "d": 0,
    "e": 0,
    "f": 0,
    "g": 0,
    "h": 0,
    "i": 0,
    "j": 0,
    "k": 0,
    "l": 0,
    "m": 0,
    "n": 0,
    "o": 0,
    "p": 0,
    "q": 0,
    "r": 0,
    "s": 1,
    "t": 0,
    "u": 0,
    "v": 0,
    "w": 0,
    "x": 0,
    "y": 0,
    "z": 0,
    "←": 0,
    "↑": 0,
    "→": 0,
    "↩": 0,
  },
  "lack": 1,
  "lackCounter": Map {
    "也" => 1,
  },
  "littleFD": 0,
  "longFD": 0,
  "multiSpan": 0,
  "sameFingers": 2,
  "singleSpan": 0,
  "trible": 0,
  "words": 2,
  "wordsChar": 0,
  "wordsDist": [
    7,
    0,
    0,
  ],
}
`;
